cluster/distributedtask/manager.go

```diff
@@ -0,0 +1,229 @@
+		}
+
+		if seqNum <= task.Version {
+			return errors.Errorf("task %s/%s is already finished with version %d", r.Namespace, r.Id, task.Version)
+		}
+	}
+
+	m.setTaskWithLock(&Task{
+		Namespace:      r.Namespace,
+		TaskDescriptor: TaskDescriptor{ID: r.Id, Version: seqNum},
+		Payload:        r.Payload,
+		Status:         TaskStatusStarted,
+		StartedAt:      time.UnixMilli(r.SubmittedAtUnixMillis),
+		FinishedNodes:  map[string]bool{},
+	})
+
+	return nil
+}
+
+func (m *Manager) RecordNodeCompletion(c *api.ApplyRequest, numberOfNodesInTheCluster int) error {
+	var r api.RecordDistributedTaskNodeCompletionRequest
+	if err := json.Unmarshal(c.SubCommand, &r); err != nil {
+		return errors.Wrap(err, "unmarshal record task node completion request")
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	task, err := m.findVersionedTaskWithLock(r.Namespace, r.Id, r.Version)
+	if err != nil {
+		return err
+	}
+
+	if task.Status != TaskStatusStarted {
+		return errors.Errorf("task %s/%s/%d is no longer running", r.Namespace, r.Id, task.Version)
+	}
+
+	if r.Error != nil {
+		task.Status = TaskStatusFailed
+		task.Error = *r.Error
+		task.FinishedAt = time.UnixMilli(r.FinishedAtUnixMillis)
+		return nil
+	}
+
+	task.FinishedNodes[r.NodeId] = true
+	if len(task.FinishedNodes) == numberOfNodesInTheCluster {
+		task.Status = TaskStatusFinished
+		task.FinishedAt = time.UnixMilli(r.FinishedAtUnixMillis)
+		return nil
+	}
+
+	return nil
+}
+
+func (m *Manager) CancelTask(a *api.ApplyRequest) error {
+	var r api.CancelDistributedTaskRequest
+	if err := json.Unmarshal(a.SubCommand, &r); err != nil {
+		return errors.Wrap(err, "unmarshal cancel task request")
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	task, err := m.findVersionedTaskWithLock(r.Namespace, r.Id, r.Version)
+	if err != nil {
+		return err
+	}
+
+	if task.Status != TaskStatusStarted {
+		return errors.Errorf("task %s/%s/%d is no longer running", r.Namespace, r.Id, task.Version)
+	}
+
+	task.Status = TaskStatusCancelled
+	task.FinishedAt = time.UnixMilli(r.CancelledAtUnixMillis)
+	return nil
+}
+
+func (m *Manager) CleanUpTask(a *api.ApplyRequest) error {
+	var r api.CleanUpDistributedTaskRequest
+	if err := json.Unmarshal(a.SubCommand, &r); err != nil {
+		return errors.Wrap(err, "unmarshal clean up task request")
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	task, err := m.findVersionedTaskWithLock(r.Namespace, r.Id, r.Version)
+	if err != nil {
+		return err
+	}
+
+	if task.Status == TaskStatusStarted {
+		return errors.Errorf("task %s/%s/%d is still running", r.Namespace, r.Id, task.Version)
+	}
+
+	if m.clock.Since(task.FinishedAt) <= m.completedTaskTTL {
+		return errors.Errorf("task %s/%s/%d is too fresh to clean up", r.Namespace, r.Id, task.Version)
+	}
+
+	delete(m.tasks[task.Namespace], task.ID)
+	return nil
+}
+
+func (m *Manager) ListDistributedTasks(_ context.Context) (map[string][]*Task, error) {
```

Comment:
question: if this method doesn't require `context` can we just remove this argument?

Reply:
I wanted for it to match a method signature of Raft:
```golang
func (s *Raft) ListDistributedTasks(ctx context.Context) (map[string][]*distributedtask.Task, error) {
	command := &cmd.QueryRequest{
		Type: cmd.QueryRequest_TYPE_DISTRIBUTED_TASK_LIST,
	}
	queryResp, err := s.Query(ctx, command)
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %w", err)
	}

	response := distributedtask.ListDistributedTasksResponse{}
	if err = json.Unmarshal(queryResp.Payload, &response); err != nil {
		return nil, fmt.Errorf("failed to unmarshal query result: %w", err)
	}

	return response.Tasks, nil
}
```